tests/ignite/contrib/metrics/test_average_precision.py

```diff
@@ -108,4 +148,100 @@ def update_fn(engine, batch):
     data = list(range(size // batch_size))
     ap = engine.run(data, max_epochs=1).metrics["ap"]
 
-    assert ap == np_ap
+    assert ap == pytest.approx(np_ap)
+
+
+def _test_distrib_compute(device):
```

Comment:
Integration test with Engine would be more interesting, I think. A minor testing issue here is that we perform only 1 update, but typically we do a lot of updates. This should be tested. @KickItLikeShika please take a look how we do tests for other core metrics and try to bake an integration test. You can indeed test that with CPU, please see CONTRIBUTING, I added a small section about how to run a distributed test. The idea is to define an Engine, attach the metric and feed in some random data which should be also different for every rank.

Reply:
What about updating the tests of `CohenKappa` the same way too? I will make tests similar to these here: [Accuracy](https://github.com/pytorch/ignite/blob/master/tests/ignite/metrics/test_accuracy.py), [Confusion Matrix](https://github.com/pytorch/ignite/blob/master/tests/ignite/metrics/test_confusion_matrix.py)

Reply:
Yes, CohenKappa should be also updated to more general testing way.

> I will make tests similar to these here: Accuracy, Confusion Matrix

Sounds good !

Reply:
Hey @vfdev-5, obviously i'm using `pytest.parametrize()` to test all weighting types in `CohenKappa`. Can i do that for distributed tests for test all the weighting types? I don't actually know but i tried that and it seemed working. Here is what i have done:
```
def _test_distirb_binary_input_N(device, weights):
    rank = idist.get_rank()
    
    def _test(metric_device):
        metric_device = torch.device(metric_device)
        ck = CohenKappa(device=metric_device, weights=weights)
        ..........
        ..........

    for _ in range(3):
        _test("cpu")
        if device.type != "xla":
            _test(idist.device())

@pytest.mark.distributed
@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
@pytest.mark.parametrize("weights", [None, "linear", "quadratic"])
def test_distrib_cpu(distributed_context_single_node_gloo, weights):

    device = torch.device("cpu")
    _test_distirb_binary_input_N(device, weights)

```
What do you think?

Reply:
Looks great ! I do not know either if it can work but I think yes. What logs look like when you launch it with pytest ?

Reply:
Nothing suspicious, it works fine!

Reply:
I mean can you put an extract of it here with the tests related to CohenKappa

Reply:
```
<frozen importlib._bootstrap>:219
<frozen importlib._bootstrap>:219
<frozen importlib._bootstrap>:219
<frozen importlib._bootstrap>:219
  <frozen importlib._bootstrap>:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
```

Reply:
well, this does not help to understand :)
You can push your code to the PR and we can see how it works on the CI.

Reply:
@vfdev-5, i'm almost done, all the tests work fine but when running the tests with CPU as provided in CONTRIBUTING i get this error:
```
.......
tests/ignite/contrib/metrics/test_precision_recall_curve.py:277: in _test
    precision, recall, thresholds = prc.compute()
ignite/metrics/epoch_metric.py:145: in compute
    result = cast(float, idist.broadcast(result, src=0))
ignite/distributed/utils.py:405: in broadcast
    return _model.broadcast(tensor, src=src)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ignite.distributed.comp_models.native._NativeDistModel object at 0x7f6c69bf12e0>
tensor = (array([0.35      , 0.31578947, 0.33333333, 0.35294118, 0.375     ,
       0.33333333, 0.35714286, 0.30769231, 0.33333...434 , 0.61502165, 0.6177955 ,
       0.69359916, 0.69379395, 0.92483944, 0.9693086 , 0.9856615 ],
      dtype=float32))
src = 0

    def broadcast(self, tensor: Union[torch.Tensor, float, str], src: int = 0) -> Union[torch.Tensor, float, str]:
        if not isinstance(tensor, (torch.Tensor, Number, str)):
>           raise TypeError(f"Unhandled input type {type(tensor)}")
E           TypeError: Unhandled input type <class 'tuple'>

ignite/distributed/comp_models/base.py:174: TypeError

```
I have spent the past few hours searching this error but got nothing. And here is the code:
```

def _test_distirb_binary_input_N(device):
    rank = idist.get_rank()
    torch.manual_seed(12)
    
    def _test(metric_device):
        metric_device = torch.device(metric_device)
        prc = PrecisionRecallCurve(device=metric_device)

        torch.manual_seed(10 + rank)
        
        y_pred = torch.rand((10, ), device=device)
        y = torch.randint(0, 2, size=(10,), device=device).long()
        prc.update((y_pred, y))
        
        # gather y_pred, y
        y_pred = idist.all_gather(y_pred)
        y = idist.all_gather(y)

        np_y = y.cpu().numpy()
        np_y_pred = y_pred.cpu().numpy()

        sk_precision, sk_recall, sk_thresholds = precision_recall_curve(np_y, np_y_pred)
        precision, recall, thresholds = prc.compute()
        
        assert np.array_equal(precision, sk_precision)
        assert np.array_equal(recall, sk_recall)
        np.testing.assert_array_almost_equal(thresholds, sk_thresholds)
        ............
        ............
    for _ in range(3):
        _test("cpu")
        if device.type != "xla":
            _test(idist.device())

@pytest.mark.distributed
@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
def test_distrib_cpu(distributed_context_single_node_gloo):

    device = torch.device("cpu")
    _test_distirb_binary_input_N(device)

```

Reply:
@KickItLikeShika the issue you encounter is with unsupported data type passed to our `broadcast` function as the error states:
```
        if not isinstance(tensor, (torch.Tensor, Number, str)):
>           raise TypeError(f"Unhandled input type {type(tensor)}")
E           TypeError: Unhandled input type <class 'tuple'>
```
normally, we should broadcast tensors, numbers or strings. But somehow a tuple is passed.

Broad cast happens in EpochMetric: 
```
ignite/metrics/epoch_metric.py:145: in compute
    result = cast(float, idist.broadcast(result, src=0))
```
on the result of self.compute_fn:
https://github.com/pytorch/ignite/blob/e6950b1f3a1f81762f31014b6741ed2c679860ba/ignite/metrics/epoch_metric.py#L141

As we say in the docs, the output of the function should be a scalar. Probably, you return something else in the function when compute PrecisionRecallCurve

Reply:
I can't really find a good and easy to use for the user, but i got another idea, `PrecisionRecallCurve` returns 3 arrays/tensors(precision, recall, thresholds), What about add an extra arg in the `PrecisionRecallCurve` to determine what you user wants to return `preicsion` or `recall` or `thresholds`. If you think this is a good idea, should i implement 3 functions for each return type? implementing it similar to `CohenKappa` and wieghts arg (BC-breaking)?

Reply:
Oh, I see better what is the issue: sklearn's precision_recall_curve returns a tuple and not a scalar. Thanks for pointing out that !

I think, what can be done is the following, let's improve EpochMetric to support various outputs. What can be done:
- add to EpochMetric: additional check that result is not None after this code
https://github.com/pytorch/ignite/blob/e6950b1f3a1f81762f31014b6741ed2c679860ba/ignite/metrics/epoch_metric.py#L141

- try to see if we could use [apply_to_type](https://github.com/pytorch/ignite/blob/e6950b1f3a1f81762f31014b6741ed2c679860ba/ignite/utils.py#L45) function to apply `idist.broadcast` on the result: `apply_to_type(result, (torch.Tensor, Number, str), partial(idist.broadcast, src=0))` ?

Reply:
@vfdev-5 `apply_to_type` raised a `TypeError`
```
E       TypeError: x must contain (<class 'torch.Tensor'>, <class 'float'>, <class 'str'>), dicts or lists; found <class 'numpy.ndarray'>
```
The updated `EpocMetric`

```
from functools import partial
from ignite.utils import apply_to_type
.........
.........
if idist.get_rank() == 0:
            # Run compute_fn on zero rank only
            result = self.compute_fn(_prediction_tensor, _target_tensor)
            if result is not None:
                apply_to_type(result, (torch.Tensor, float, str), partial(idist.broadcast, src=0))
.......
```

Reply:
We can return tensors from our compute_fn in Precision Recall Curve ?

Reply:
I proposed to return the user chooses to pass as arg `precision` or `recall` or `thresholds` and make 3 functions each of them returns something the user wants. One extra thing, whatever we do of `PrecisionRecallCurve` we will do it for `RocCurve` because it aslo return fpr, tpr, thresholds.

Reply:
But we can return tuple of tensors from compute_fn? if that's what you're asking i think yes

Reply:
Yes, let's return a tuple of tensors

Reply:
I tried this and everything now works fine! thank you sir!
and here is the implementation for `PreicsionRecallCurve` and `RocCurve` is similar
```
def precision_recall_curve_compute_fn(y_preds: torch.Tensor, y_targets: torch.Tensor) -> Union[Any, Any, Any]:
    from sklearn.metrics import precision_recall_curve

    y_true = y_targets.cpu().numpy()
    y_pred = y_preds.cpu().numpy()
    precision, recall, thresholds = precision_recall_curve(y_true, y_pred)
    precision, recall, thresholds = torch.from_numpy(precision.copy()), torch.from_numpy(recall.copy()), torch.from_numpy(thresholds.copy())
    return precision, recall, thresholds
```
The reason for using copy is sometimes the strides are negative then we will get this error:
```
ValueError: At least one stride in the given numpy array is negative, and tensors with negative strides are not currently supported. (You can probably work around this by making a copy of your array  with array.copy().)
```

Reply:
@vfdev-5, i have committed the new changed but it didn't pass the CI (expected), just to leave it for review and your tips to pass the CI, Thanks

Reply:
@vfdev-5 there are some `RuntimeErrors` please check it and tell me how can i fix that